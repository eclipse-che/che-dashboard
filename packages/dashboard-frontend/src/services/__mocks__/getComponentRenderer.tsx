/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

export default function <T extends Array<unknown>>(
  getComponent: (..._args: T) => React.ReactElement,
): {
  createSnapshot: (...args: T) => renderer.ReactTestRenderer;
  renderComponent: (...args: T) => {
    reRenderComponent: (..._args: T) => void;
  };
} {
  return {
    createSnapshot: (...args: T) => {
      return renderer.create(getComponent(...args));
    },
    renderComponent: (...args: T) => {
      const res = render(getComponent(...args));
      return {
        reRenderComponent: (..._args: T) => {
          res.rerender(getComponent(..._args));
        },
      };
    },
  };
}
