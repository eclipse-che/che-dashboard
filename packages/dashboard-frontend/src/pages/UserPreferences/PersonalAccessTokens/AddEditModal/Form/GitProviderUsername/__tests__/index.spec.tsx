/*
 * Copyright (c) 2018-2023 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Form } from '@patternfly/react-core';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { GitProviderUsername } from '..';
import getComponentRenderer, {
  screen,
} from '../../../../../../../services/__mocks__/getComponentRenderer';

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

const mockOnChange = jest.fn();

describe('GitProviderUsername', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('snapshot w/o username', () => {
    const snapshot = createSnapshot();
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test('snapshot with username', () => {
    const snapshot = createSnapshot('username');
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  it('should handle a correct username', () => {
    const username = 'username';
    renderComponent();

    expect(mockOnChange).not.toHaveBeenCalled();

    const input = screen.getByRole('textbox');
    userEvent.paste(input, username);

    expect(mockOnChange).toHaveBeenCalledWith(username, true);
    expect(screen.queryByText('This field is required.')).toBeFalsy();
    expect(screen.queryByText(/^The Provider Username is too long./)).toBeFalsy();
  });

  it('should handle a too long username value', () => {
    // make it long enough to be invalid
    const username = 'username'.repeat(50);
    renderComponent();

    expect(mockOnChange).not.toHaveBeenCalled();

    const input = screen.getByRole('textbox');
    userEvent.paste(input, username);

    expect(mockOnChange).toHaveBeenCalledWith(username, false);
    expect(screen.queryByText('This field is required.')).toBeFalsy();
    expect(screen.queryByText(/^The Provider Username is too long./)).toBeTruthy();
  });

  it('should handle an empty value', () => {
    const username = 'username';
    renderComponent(username);

    expect(mockOnChange).not.toHaveBeenCalled();

    const input = screen.getByRole('textbox');
    userEvent.clear(input);

    expect(mockOnChange).toHaveBeenCalledWith('', false);
    expect(screen.queryByText('This field is required.')).toBeTruthy();
    expect(screen.queryByText(/^The Provider Username is too long./)).toBeFalsy();
  });
});

function getComponent(providerUsername?: string): React.ReactElement {
  return (
    <Form>
      <GitProviderUsername providerUsername={providerUsername} onChange={mockOnChange} />
    </Form>
  );
}
