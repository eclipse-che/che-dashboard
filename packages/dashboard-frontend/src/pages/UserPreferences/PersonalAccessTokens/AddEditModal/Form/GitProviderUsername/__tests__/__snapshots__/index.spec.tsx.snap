// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitProviderUsername snapshot w/o username 1`] = `
<form
  className="pf-c-form"
  noValidate={true}
>
  <div
    className="pf-c-form__group"
  >
    <div
      className="pf-c-form__group-label"
    >
      <label
        className="pf-c-form__label"
        htmlFor="git-provider-username-label"
      >
        <span
          className="pf-c-form__label-text"
        >
          Git Provider Username
        </span>
        <span
          aria-hidden="true"
          className="pf-c-form__label-required"
        >
           
          *
        </span>
      </label>
       
    </div>
    <div
      className="pf-c-form__group-control"
    >
      <input
        aria-describedby="git-provider-username-label"
        aria-invalid={false}
        aria-label="Git Provider Username"
        className="pf-c-form-control"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Enter a Git Provider Username"
        readOnly={false}
        required={false}
        type="text"
        value=""
      />
      
    </div>
  </div>
</form>
`;

exports[`GitProviderUsername snapshot with username 1`] = `
<form
  className="pf-c-form"
  noValidate={true}
>
  <div
    className="pf-c-form__group"
  >
    <div
      className="pf-c-form__group-label"
    >
      <label
        className="pf-c-form__label"
        htmlFor="git-provider-username-label"
      >
        <span
          className="pf-c-form__label-text"
        >
          Git Provider Username
        </span>
        <span
          aria-hidden="true"
          className="pf-c-form__label-required"
        >
           
          *
        </span>
      </label>
       
    </div>
    <div
      className="pf-c-form__group-control"
    >
      <input
        aria-describedby="git-provider-username-label"
        aria-invalid={false}
        aria-label="Git Provider Username"
        className="pf-c-form-control"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Enter a Git Provider Username"
        readOnly={false}
        required={false}
        type="text"
        value="username"
      />
      
    </div>
  </div>
</form>
`;
